### PR TITLE
Add Reactor Netty version of secure chat server example reactor#761

### PR DESCRIPTION
**Pull request by jchenga**
Provide a corresponding Reactor Netty version of Secure chat server example as found in Netty project.

Related to #761

### CODE REVIEW

**Comment by jchenga** [comment]
I am not sure why build (windows-2019, nio) failed.  the check failed with assertion exception at  ITTracingHttpClientDecoratorTest, but the same test succeeded with other OS platforms.

File: reactor-netty-examples/src/main/java/reactor/netty/examples/tcp/securechat/SecureChatClient.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+				.doOnConnected(connection -> {
+					connection.addHandlerLast(new DelimiterBasedFrameDecoder(8192, Delimiters.lineDelimiter()));
+				}).secure(spec -> spec.sslContext(tcpSslContextSpec));
+
+		Connection conn = client.connectNow();
+		conn.inbound()
+				.receive()
+				.asString()
+				.doOnNext(msg -> System.out.println(msg))
+				.subscribe();
```
Comment (pderop) [inline reactor-netty-examples/src/main/java/reactor/netty/examples/tcp/securechat/SecureChatClient.java:53]:
```suggestion
				.subscribe(System.out::println);
```

what do you think about replacing doOnNext by one single subscribe method ?

**Review by pderop — CHANGES_REQUESTED** [review]
Hi @jchenga ,

thanks a lot for submitting this PR !
Now, I have made few comments, can you please check the following ?

File: reactor-netty-examples/src/main/java/reactor/netty/examples/tcp/securechat/SecureChatServer.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+										if (conn == current) {
+											sink.next(text);
+										}
+										else {
+											String msg = "[" + conn.channel().remoteAddress() + "] " + text + '\n';
+
+											conn.outbound().sendString(Mono.just(msg)).then().subscribe();
+										}
+									}
+								});
```
Comment (pderop) [inline reactor-netty-examples/src/main/java/reactor/netty/examples/tcp/securechat/SecureChatServer.java:94]:
```suggestion
                                in.withConnection(current -> Flux.fromStream(conns.values()
                                                .stream()
                                                .filter(conn -> conn != current))
                                        .map(conn -> conn.outbound()
                                                .sendString(Mono.just("[" + conn.channel().remoteAddress() + "] " + text + '\n')).then())
                                        .doOnComplete(() -> sink.next(text))
                                        .subscribe()
                                );
```

in order to encourage back pressure, what about waiting for the flux of forwarded messages to be completed before returning the message to the original user, something that would look like above ?

Comment (pderop) [reply reactor-netty-examples/src/main/java/reactor/netty/examples/tcp/securechat/SecureChatServer.java:94]: I just changed the above proposed change, please take care (I replaced map by flatMap line 86).
thanks.

Comment (pderop) [reply reactor-netty-examples/src/main/java/reactor/netty/examples/tcp/securechat/SecureChatServer.java:94]: mmm, my suggestion does not seem to work, please ignore this one for the moment. sorry.

Comment (pderop) [reply reactor-netty-examples/src/main/java/reactor/netty/examples/tcp/securechat/SecureChatServer.java:94]: `handle` is using a `SynchronousSInk`, that's why the above was not working.

File: reactor-netty-examples/src/main/java/reactor/netty/examples/tcp/securechat/SecureChatClient.java (hunk 1 above)
Comment (jchenga) [reply reactor-netty-examples/src/main/java/reactor/netty/examples/tcp/securechat/SecureChatClient.java]: @pderop thank you for the suggestion. I've made the suggested change.

File: reactor-netty-examples/src/main/java/reactor/netty/examples/tcp/securechat/SecureChatServer.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+					conns.put(id, connection);
+					connection.addHandlerLast(new DelimiterBasedFrameDecoder(8192, Delimiters.lineDelimiter()));
+				}).childObserve((connection, newState) -> {
+					if (newState == State.DISCONNECTING) {
+						// when a connection is disconnecting,
+						// remove it from the map. We don't want to hold onto
+						// a disconnected connection.
+						ChannelId id = connection.channel().id();
+						conns.remove(id);
+					}
```
[the remark this answers is not in the document]
Comment (jchenga) [reply reactor-netty-examples/src/main/java/reactor/netty/examples/tcp/securechat/SecureChatServer.java]: I've also implemented the suggested change. the code is definitely neater. thanks!

**Review by pderop — APPROVED** [review]
thanks @jchenga,

+1

@violetagg , please take a look

File: reactor-netty-examples/src/main/java/reactor/netty/examples/tcp/securechat/SecureChatClient.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+					connection.addHandlerLast(new DelimiterBasedFrameDecoder(8192, Delimiters.lineDelimiter()));
+				}).secure(spec -> spec.sslContext(tcpSslContextSpec));
+
+		Connection conn = client.connectNow();
+		conn.inbound()
+				.receive()
+				.asString()
+				.subscribe(System.out::println);
+
+		Scanner scanner = new Scanner(System.in);
```
Comment (violetagg) [inline reactor-netty-examples/src/main/java/reactor/netty/examples/tcp/securechat/SecureChatClient.java:54]:
```suggestion
		Scanner scanner = new Scanner(System.in, StandardCharsets.UTF_8.name());
```

**Review by violetagg — CHANGES_REQUESTED** [review]
nice!

File: reactor-netty-examples/src/main/java/reactor/netty/examples/tcp/securechat/SecureChatClient.java (hunk 4 above)
Comment (jchenga) [reply reactor-netty-examples/src/main/java/reactor/netty/examples/tcp/securechat/SecureChatClient.java]: @violetagg i've made the change you requested. please have a look. thanks!

**Review by violetagg — APPROVED** [review]

**Comment by pderop** [comment]
@jchenga , thanks for the PR;
@violetagg , thanks for the review;  I have merged it into main and netty5 branches.

**Pull request merged**